File: ella/week7/2630.py
# 1780.py랑 똑같이 풀었는데 이건 맞음.. 다른점이 뭐지??
import sys

input = sys.stdin.readline
n = int(input())

p = [list(map(int, input().split())) for _ in range(n)]
answer = [0, 0]

def dev(x, y, z):
    global answer
    color = p[x][y]
    for i in range(x, x+z):
        for j in range(y, y+z):
            if color != p[i][j]:
                for a in range(2):
                    for b in range(2):
                        dev(x + (z//2*a), y + (z//2*b), z//2)
                return
    if color == 0:
        answer[0] += 1
    elif color != 0:
        answer[1] += 1

dev(0, 0, n)
print(answer[0])
print(answer[1])

File: leo/week7/2630.js
// Run by Node.js
const readline = require("readline");

(async () => {
  let rl = readline.createInterface({ input: process.stdin });

  let input = [];
  for await (const line of rl) {
    input.push(line);
  }

  let N = parseInt(input[0]);
  let paper = [];
  for (let i = 1; i <= N; i++) {
    paper.push(input[i].trim().split(" ").map(Number));
  }

  let counts = { 0: 0, 1: 0 };

  function checkSameColor(x, y, size) {
    let color = paper[x][y];
    for (let i = x; i < x + size; i++) {
      for (let j = y; j < y + size; j++) {
        if (paper[i][j] !== color) {
          return false;
        }
      }
    }
    return true;
  }

  function divideAndConquer(x, y, size) {
    if (checkSameColor(x, y, size)) {
      counts[paper[x][y]]++;
    } else {
      let newSize = size / 2;
      divideAndConquer(x, y, newSize); // Top-left
      divideAndConquer(x, y + newSize, newSize); // Top-right
      divideAndConquer(x + newSize, y, newSize); // Bottom-left
      divideAndConquer(x + newSize, y + newSize, newSize); // Bottom-right
    }
  }

  divideAndConquer(0, 0, N);

  console.log(counts["0"]);
  console.log(counts["1"]);

  process.exit();
})();
